swift: field of type `self` conflicts with implicit `self` in initializer

When generating Swift code where an object has a field named `self`, this synthesizes code that looks like

```swift
public struct Foo: GraphQLSelectionSet {
    // …
    public init(`self`: `Self`? = nil) {
        self.init(unsafeResultMap: ["__typename": "Foo", "self": `self`.flatMap { … }])
    }
    // …
}
```

Unfortunately, while the escaping makes this syntactically correct, the parameter <code>\`self\`</code> conflicts with the implicit parameter `self` because, while `self` is indeed a keyword, the compiler also treats it as equivalent to the identifier.

The use of the synthesized variable <code>public var \`self\`: \`Self\`</code> is perfectly fine as it doesn't actually reference the identifier within itself, and AFAICT no other generated code in the type references the `self` var (if it did, <code>self.\`self\`</code> would work fine of course). So we really just need to fix the initializer.

I'm pretty sure `self` is the only token we have to treat specially here, and I'm tempted to just name it `_self`, although of course if there's another field explicitly named `_self` then it would conflict; maybe we can detect that and tack on trailing underscores until we find something unique, e.g. `_self_`, `_self__`, etc. In any case, this change wouldn't be visible outside of the function as long as we keep the external parameter name as <code>\`self\`</code>. With the proposed fix the above type would become

```swift
public struct Foo: GraphQLSelectionSet {
    // …
    public init(`self` _self: `Self`? = nil) {
        self.init(unsafeResultMap: ["__typename": "Foo", "self": _self.flatMap { … }])
    }
    // …
}
```

CC @designatednerd

Just tested this again, `npm install --clean` from the root of `apollo-tooling`, and then ran `packages/apollo/bin/run help client:codegen` and I get a bunch of spew about `Error Plugin: apollo: Cannot find module 'apollo-env'` followed by `Error: command client:codegen not found`

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,6 +9,7 @@
 - `apollo-codegen-scala`
   - <First `apollo-codegen-scala` related entry goes here>
 - `apollo-codegen-swift`
+  - Ensure fields named `self` don't cause compilation errors in the generated code [#1533](https://github.com/apollographql/apollo-tooling/pull/1533)
   - Preserve leading/trailing underscores on field names [#1533](https://github.com/apollographql/apollo-tooling/pull/1533)
 - `apollo-codegen-typescript`
   - <First `apollo-codegen-typescript` related entry goes here>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -9,7 +9,7 @@
 - `apollo-codegen-scala`
   - <First `apollo-codegen-scala` related entry goes here>
 - `apollo-codegen-swift`
-  - <First `apollo-codegen-swift` related entry goes here>
+  - Preserve leading/trailing underscores on field names [#1533](https://github.com/apollographql/apollo-tooling/pull/1533)
 - `apollo-codegen-typescript`
   - <First `apollo-codegen-typescript` related entry goes here>
 - `apollo-env`

--- a/packages/apollo-codegen-swift/src/__tests__/codeGeneration.ts
+++ b/packages/apollo-codegen-swift/src/__tests__/codeGeneration.ts
@@ -455,6 +455,30 @@ describe("Swift code generation", () => {
       expect(generator.output).toMatchSnapshot();
     });
 
+    it(`should preserve leading and trailing underscores on fields`, () => {
+      const { operations } = compile(`
+        query Hero {
+          hero {
+            _name: name
+            _camel_case_id__: id
+          }
+        }
+      `);
+
+      const selectionSet = (operations["Hero"].selectionSet
+        .selections[0] as Field).selectionSet as SelectionSet;
+
+      generator.structDeclarationForSelectionSet(
+        {
+          structName: "Hero",
+          selectionSet
+        },
+        false
+      );
+
+      expect(generator.output).toMatchSnapshot();
+    });
+
     it(`should escape reserved keywords in a struct declaration for a selection set`, () => {
       const { operations } = compile(`
         query Hero {

--- a/packages/apollo-codegen-swift/src/__tests__/codeGeneration.ts
+++ b/packages/apollo-codegen-swift/src/__tests__/codeGeneration.ts
@@ -505,6 +505,46 @@ describe("Swift code generation", () => {
       expect(generator.output).toMatchSnapshot();
     });
 
+    it(`should escape init specially in a struct declaration initializer for a selection set`, () => {
+      const { operations } = compile(`
+        query Humans {
+          human(id: 0) {
+            self: friends {
+              id
+            }
+          }
+          human(id: 1) {
+            self: friends {
+              id
+            }
+            _self: name
+          }
+        }
+      `);
+
+      const human0 = (operations["Humans"].selectionSet.selections[0] as Field)
+        .selectionSet as SelectionSet;
+      const human1 = (operations["Humans"].selectionSet.selections[1] as Field)
+        .selectionSet as SelectionSet;
+
+      generator.structDeclarationForSelectionSet(
+        {
+          structName: "Human",
+          selectionSet: human0
+        },
+        false
+      );
+      generator.structDeclarationForSelectionSet(
+        {
+          structName: "Human",
+          selectionSet: human1
+        },
+        false
+      );
+
+      expect(generator.output).toMatchSnapshot();
+    });
+
     it(`should generate a nested struct declaration for a selection set with subselections`, () => {
       const { operations } = compile(`
         query Hero {

--- a/packages/apollo-codegen-swift/src/codeGeneration.ts
+++ b/packages/apollo-codegen-swift/src/codeGeneration.ts
@@ -643,7 +643,9 @@ export class SwiftAPIGenerator extends SwiftGenerator<CompilerContext> {
                 swift`"__typename": ${SwiftSource.string(
                   variant.possibleTypes[0].toString()
                 )}`,
-                ...properties.map(this.propertyAssignmentForField, this)
+                ...properties.map(p =>
+                  this.propertyAssignmentForField(p, properties)
+                )
               ],
               ", "
             ) || ":",
@@ -683,7 +685,9 @@ export class SwiftAPIGenerator extends SwiftGenerator<CompilerContext> {
                   swift`"__typename": ${SwiftSource.string(
                     possibleType.toString()
                   )}`,
-                  ...properties.map(this.propertyAssignmentForField, this)
+                  ...properties.map(p =>
+                    this.propertyAssignmentForField(p, properties)
+                  )
                 ],
                 ", "
               ) || ":",
@@ -695,13 +699,16 @@ export class SwiftAPIGenerator extends SwiftGenerator<CompilerContext> {
     }
   }
 
-  propertyAssignmentForField(field: {
-    responseKey: string;
-    propertyName: string;
-    type: GraphQLType;
-    isConditional?: boolean;
-    structName?: string;
-  }): SwiftSource {
+  propertyAssignmentForField(
+    field: {
+      responseKey: string;
+      propertyName: string;
+      type: GraphQLType;
+      isConditional?: boolean;
+      structName?: string;
+    },
+    properties: { propertyName: string }[]
+  ): SwiftSource {
     const {
       responseKey,
       propertyName,
@@ -709,16 +716,20 @@ export class SwiftAPIGenerator extends SwiftGenerator<CompilerContext> {
       isConditional,
       structName
     } = field;
+    const parameterName = this.helpers.internalParameterName(
+      propertyName,
+      properties
+    );
     const valueExpression = isCompositeType(getNamedType(type))
       ? this.helpers.mapExpressionForType(
           type,
           isConditional,
           expression => swift`${expression}.resultMap`,
-          SwiftSource.identifier(propertyName),
+          SwiftSource.identifier(parameterName),
           structName!,
           "ResultMap"
         )
-      : SwiftSource.identifier(propertyName);
+      : SwiftSource.identifier(parameterName);
     return swift`${SwiftSource.string(responseKey)}: ${valueExpression}`;
   }
 
@@ -882,7 +893,12 @@ export class SwiftAPIGenerator extends SwiftGenerator<CompilerContext> {
 
     this.withinBlock(() => {
       properties.forEach(({ propertyName }) => {
-        this.printOnNewline(swift`self.${propertyName} = ${propertyName}`);
+        this.printOnNewline(
+          swift`self.${propertyName} = ${this.helpers.internalParameterName(
+            propertyName,
+            properties
+          )}`
+        );
       });
     });
   }
@@ -891,12 +907,20 @@ export class SwiftAPIGenerator extends SwiftGenerator<CompilerContext> {
     this.print(swift`(`);
     this.print(
       join(
-        properties.map(({ propertyName, typeName, isOptional }) =>
-          join([
-            swift`${propertyName}: ${typeName}`,
+        properties.map(({ propertyName, typeName, isOptional }) => {
+          const internalName = this.helpers.internalParameterName(
+            propertyName,
+            properties
+          );
+          const decl =
+            internalName === propertyName
+              ? propertyName
+              : swift`${propertyName} ${internalName}`;
+          return join([
+            swift`${decl}: ${typeName}`,
             isOptional ? swift` = nil` : undefined
-          ])
-        ),
+          ]);
+        }),
         ", "
       )
     );
@@ -1168,12 +1192,20 @@ export class SwiftAPIGenerator extends SwiftGenerator<CompilerContext> {
         this.print(swift`(`);
         this.print(
           join(
-            properties.map(({ propertyName, typeName, isOptional }) =>
-              join([
-                swift`${propertyName}: ${typeName}`,
+            properties.map(({ propertyName, typeName, isOptional }) => {
+              const internalName = this.helpers.internalParameterName(
+                propertyName,
+                properties
+              );
+              const decl =
+                internalName === propertyName
+                  ? propertyName
+                  : swift`${propertyName} ${internalName}`;
+              return join([
+                swift`${decl}: ${typeName}`,
                 isOptional ? swift` = nil` : undefined
-              ])
-            ),
+              ]);
+            }),
             ", "
           )
         );
@@ -1186,7 +1218,12 @@ export class SwiftAPIGenerator extends SwiftGenerator<CompilerContext> {
               join(
                 properties.map(
                   ({ name, propertyName }) =>
-                    swift`${SwiftSource.string(name)}: ${propertyName}`
+                    swift`${SwiftSource.string(
+                      name
+                    )}: ${this.helpers.internalParameterName(
+                      propertyName,
+                      properties
+                    )}`
                 ),
                 ", "
               ) || ":",

--- a/packages/apollo-codegen-swift/src/helpers.ts
+++ b/packages/apollo-codegen-swift/src/helpers.ts
@@ -15,7 +15,10 @@ import {
   isEnumType
 } from "graphql";
 
-import { camelCase, pascalCase } from "change-case";
+import {
+  camelCase as _camelCase,
+  pascalCase as _pascalCase
+} from "change-case";
 import * as Inflector from "inflected";
 import { join, wrap } from "apollo-codegen-core/lib/utilities/printing";
 
@@ -346,4 +349,34 @@ function makeClosureSignature(
   }
   closureSignature.append(swift` in`);
   return closureSignature;
+}
+
+/**
+ * Converts a value from "underscore_case" to "camelCase".
+ *
+ * This preserves any leading/trailing underscores.
+ */
+function camelCase(value: string): string {
+  const [_, prefix, middle, suffix] = value.match(/^(_*)(.*?)(_*)$/) || [
+    "",
+    "",
+    value,
+    ""
+  ];
+  return `${prefix}${_camelCase(middle)}${suffix}`;
+}
+
+/**
+ * Converts a value from "underscore_case" to "PascalCase".
+ *
+ * This preserves any leading/trailing underscores.
+ */
+function pascalCase(value: string): string {
+  const [_, prefix, middle, suffix] = value.match(/^(_*)(.*?)(_*)$/) || [
+    "",
+    "",
+    value,
+    ""
+  ];
+  return `${prefix}${_pascalCase(middle)}${suffix}`;
 }

--- a/packages/apollo-codegen-swift/src/helpers.ts
+++ b/packages/apollo-codegen-swift/src/helpers.ts
@@ -135,6 +135,24 @@ export class Helpers {
     );
   }
 
+  /**
+   * Returns the internal parameter name for a given property name.
+   *
+   * If the property name is valid to use, it's returned directly. Otherwise it's prefixed with an
+   * underscore and modified until it's unique among the given property set.
+   * @param propertyName The name of the property.
+   * @param properties A list of properties that should be consulted when producing a unique name.
+   * @returns The name to use for the internal parameter name for the property.
+   */
+  internalParameterName(
+    propertyName: string,
+    properties: { propertyName: string }[]
+  ): string {
+    return SwiftSource.isValidParameterName(propertyName)
+      ? propertyName
+      : makeUniqueName(`_${propertyName}`, properties);
+  }
+
   // Properties
 
   propertyFromField(
@@ -349,6 +367,25 @@ function makeClosureSignature(
   }
   closureSignature.append(swift` in`);
   return closureSignature;
+}
+
+/**
+ * Takes a proposed name and modifies it to be unique given a list of properties.
+ * @param proposedName The proposed name that shouldn't conflict with any property.
+ * @param properties A list of properties the name shouldn't conflict with.
+ * @returns A name based on `proposedName` that doesn't match any existing property name.
+ */
+function makeUniqueName(
+  proposedName: string,
+  properties: { propertyName: string }[]
+): string {
+  // Assume conflicts are very rare and property lists are short, and just do a linear search. If
+  // we find a conflict, start over with the modified name.
+  for (let name = proposedName; ; name += "_") {
+    if (properties.every(prop => prop.propertyName != name)) {
+      return name;
+    }
+  }
 }
 
 /**

--- a/packages/apollo-codegen-swift/src/language.ts
+++ b/packages/apollo-codegen-swift/src/language.ts
@@ -216,6 +216,21 @@ export class SwiftSource {
   }
 
   /**
+   * Returns whether the given name is valid as a method parameter name.
+   *
+   * Certain tokens aren't valid as method parameter names, even when escaped with backticks, as
+   * the compiler interprets the keyword and identifier as the same thing. In particular, `self`
+   * works this way.
+   * @param input The proposed parameter name.
+   * @returns `true` if the name can be used, or `false` if it needs a separate internal parameter
+   * name.
+   */
+  static isValidParameterName(input: string): boolean {
+    // Right now `self` is the only known token that we can't use with escaping.
+    return input !== "self";
+  }
+
+  /**
    * Template tag for producing a `SwiftSource` value without performing escaping.
    *
    * This is identical to evaluating the template without the tag and passing the result to `new

